박용석-자판기 만들기 프로젝트

up456 opened the pull request.

fabbro1 committed:
프로젝트 셋팅

fabbro1 committed:
feat: 불필요한 코드 삭제, css 스타일 유지

fabbro1 committed:
feat: 투입, 반환 기능을 위한 기본 UI 작성

- 자판기 총 금액판
- 돈 입력 창
- 투입, 반환 버튼
- 로그 창

fabbro1 committed:
feat: 현재 잔액 표시 기능 추가

fabbro1 committed:
feat: 투입 기능 구현

fabbro1 committed:
feat: 로그 박스 자동 스크롤 기능 추가

- 로그 업데이트 로직 분리하여 적용

fabbro1 committed:
refactor: 자판기 2개로 확장하고 Class에서 모두 처리하도록 변경

fabbro1 committed:
feat: 잔돈 반환 기능 추가

fabbro1 committed:
feat: 자판기 상품 버튼 등록 기능 추가

fabbro1 committed:
design: 상품 버튼 스타일링

fabbro1 committed:
refactor: 타입 지정

fabbro1 committed:
feat: 구매하기 기능 추가

fabbro1 committed:
feat: 구매 실패 시 상품 버튼을 누르는 동안 잔액표시창에 상품 가격 노출기능

fabbro1 committed:
refactor: 버튼 스타일을 버튼을 생성하는곳에서 지정하도록 변경

fabbro1 committed:
design: 상품 버튼 hover, active 효과 추가

fabbro1 committed:
refactor: 함수명을 보다 명확하게 수정

- 값을 변경하는 느낌보다는 화면에 보여주는 느낌으로 update에서 render로 수정

fabbro1 committed:
refactor: 현재 잔액 표시판이 아닌 금액 표시판으로 명칭을 변경 및 랜더링 함수로 통일

fabbro1 committed:
refactor: input 요소를 생성자에 추가하여 사용

fabbro1 committed:
design: 금액 표시판, 로그에서의 숫자가 3자리마다 ,를 찍도록 수정

fabbro1 committed:
design: 금액 투입 input에서도 숫자가 3자리마다 ,를 찍도록 수정

- input 타입 number -> text로 변경
- input 입력시 마다 맨앞에 0제거, 3자리마다 ,추가 로직 추가

fabbro1 committed:
feat: 로그 문구 개선

fabbro1 committed:
design: 로그 색상 부여

- 구매 성공 => 초록
- 구매 실패 => 빨강
- 그외 => 검정

fabbro1 committed:
feat: 투입 작동 후 자동 input 포커싱 기능 추가

fabbro1 committed:
feat: Enter key로 금액 입려 기능 추가

fabbro1 committed:
design: 전체 스타일링

fabbro1 committed:
feat: 자판기 추가 2개 설치

fabbro1 committed:
design: 상품이 다섯 글자를 넘어가면 말줄임 처리

fabbro1 committed:
docs: README 업데이트

fabbro1 committed:
comment: 설명 주석 수정

fabbro1 committed:
feat: 투입, 반환 버튼의 aria-label를 보다 명확하게 변경

fabbro1 committed:
fix: button안에 태그를 inline요소로 변경

taggon commented on src/VendingMachine.js:
```diff
@@ -0,0 +1,194 @@
+      button.appendChild(productPriceEl);
+
+      // 버튼 클릭 이벤트 추가
+      button.addEventListener("click", () => {
+        this.buyProduct(product);
+      });
+
+      // 잔액 부족으로 구매 실패시 버튼을 누르고 있을 때 이벤트 추가
+      button.addEventListener("mousedown", () => {
+        if (this.currentMoney < product.price) {
+          this.renderMoneyBoard(product.price);
+        }
+      });
+
+      // 잔액 부족으로 구매 실패시 버튼에서 손을 뗄 때 이벤트 추가
+      button.addEventListener("mouseup", () => {
+        this.renderMoneyBoard();
+      });
```
이벤트 리스너를 이렇게 하나하나 추가하는 것보다 더 효율적인 방법은 없을까요?

up456 replied:
이벤트 위임으로 수정했습니다~!

taggon reviewed (commented):
클래스 하나에 너무 다양한 기능이 포함되어 있는 것으로 보입니다.
로그와 투입금액을 별도 클래스나 함수 등으로 분리해보세요.

fabbro1 committed:
refactor:  Product 버튼 이벤트 처리 방식 개선

- 이벤트 위임 패턴 적용으로 반복적으로 버튼에 이벤트를 붙이는 로직 제거
- 버튼에 데이터 속성 추가로 Prodcut 정보 전달

fabbro1 committed:
refactor: 보다 명확한 변수명으로 변경

- "product-container"에서  실제 담고 있는 요소와 역할에 맞게 "product-button-list"로 변경
- setupProducts 메서드명을 더 명확하고 일관성 있게 initProducts로 변경

fabbro1 committed:
refactor: 잔액이라는 어감에 더 맞게 currentMoney단어를 currentBalance로 변경

fabbro1 committed:
refactor: 잔액 표시 함수명 개선 (renderBalanceBoard → updateBalanceDisplay)

- 함수의 목적을 더 명확하게 전달하도록 이름 변경

fabbro1 committed:
refactor: 현재 잔액 클래스를 만들어서 분리

fabbro1 committed:
refactor: 로그 클래스 만들어서 분리

- 로그 클래스를 분리하니 로그는 더하는 경우만 있다는 것을 인지하게 되었다.
- 로그가 추가될때마다 render를 신경쓰지 않아도 되어서 좋았다.

fabbro1 committed:
refactor: 변수명, 함수명 개선

- CurrentBalance클래스에서 amount를 money로 변수명 변경
- returnChanges에서 returnChange로 함수명 변경

fabbro1 committed:
fix: 구매 실패시 상품 가격이 노출되지 않는 버그 수정

fabbro1 committed:
refactor: 상품 버튼 클래스 만들어서 분리

- 핸드러까지 포함시킬려고 했으나 상품 버튼 클래스에서 다른 클래스의 컨트롤을 하는 것은 좋지 않은 패턴인 것을 배움
- 통합관리 역할을 하는 자판기 클래스에서 핸들러 전달

fabbro1 committed:
refactor: 잔액 관리 클래스 이름을 CurrentBalanceManager로 변경

- 잔액 관리 기능을 더 명확한 표현의 이름으로 변경
- 다른 관리 클래스와 일관성 유지

fabbro1 committed:
refactor: 폴더 구조 및 파일명 개선

- VendingMachine 클래스와 같은 UI단위의 클래스는 components 디렉토리로 이동
- CurrentBalanceManager, LogManager, ProductButtonManager 클래스와 같이 기능을 관리하는 클래스들은 Managers폴더로 이동
- 스타일 시트를 styles.css 에서 베이스 스타일이라는 느낌을 더 주기 위해 base.css로 파일명 변경

fabbro1 committed:
docs: README 업데이트

- 파일구조 수정
- 클래스 다이어그램 추가
- 유저 플로우 추가

fabbro1 committed:
docs: README 업데이트

- 오타 및 적절한 단어로 수정

fabbro1 committed:
docs: README 업데이트

- 클래스 다이어그램 수정

fabbro1 committed:
refactor: CurrentBalanceManager 클래스에 private 필드 추가

- balance을 private 필드로 변경

fabbro1 committed:
refactor: LogManager 클래스에 private 키워드 적용

- logs와 logsContainerEl을 private 필드로 변경
- render메서드를 private로 변경

fabbro1 committed:
refactor: private 필드와 메서드 적용

- 클래스 내부 필드들을 private으로 변경
- 내부 메서드들을 private으로 변경
- init() 메서드는 public 유지

fabbro1 committed:
refactor: VendingMachine 클래스에 private 필드와 메서드 적용

- 모든 내부 필드를 private으로 변경
- 내부에서만 사용되는 메서드들을 private으로 변경
- init() 메서드는 public 유지

fabbro1 committed:
style: 사용되지 않는 코드 삭제

fabbro1 committed:
style: 필드 순서를 성격에 맞게 분류하여 배치

fabbro1 committed:
refactor: currentBalance 다른 Manager들과 같은 패턴의 변수명으로 변경

fabbro1 committed:
style: 필드 순서를 성격에 맞게 분류하여 배치

fabbro1 committed:
docs: README 업데이트

- 클레스 변경사항에 따른 클래스 다이어그램 업데이트

fabbro1 committed:
refactor: aria-label 상수화 하여 테스트 코드에 활용 및 유지보수성 강과

fabbro1 committed:
chore: vitest 및 테스팅 라이브러리 설치

fabbro1 committed:
fix: 테스트 환경에서 scrollTo가 동작하지 않을 수 있음으로 예외처리

fabbro1 committed:
feat: test 기본환경 셋팅 및 테스트 Dom 생성 함수 추가

- createVendingMachineDOM 함수 추가
- 테스트에 필요한 UI 요소를 만든다.

fabbro1 committed:
refactor: 다른 메서드들과 일관성있게 화살표 함수로 수정

fabbro1 committed:
feat: 자판기 기본테스트문 작성

-최소 클릭으로 돈 입력되는지 확인하는 테스트 추가

fabbro1 committed:
feat: 테스트를 위한 함수 추가

fabbro1 committed:
feat: 금액투입을 했을 때 부수적인 내용 테스트 추가

fabbro1 committed:
feat: 금액 투입 로직을 함수로 분리하고 Enter키에도 동작할 수 있게 설정

fabbro1 committed:
feat: input 값 입력 및 Enter 키 이벤트에 따른 금액 투입 테스트 추가

fabbro1 committed:
refactor: 조금 더 테스트 시나리오에 맞는 테스트코드로 수정

fabbro1 committed:
feat: 테스트 Dom 생성기에 반환 버튼 추가

fabbro1 committed:
feat: 반환 기능 관련 테스트 코드 추가

fabbro1 committed:
fix: 잘못된 test 코드 삭제

fabbro1 committed:
feat: 자판기를 html에서 복제하지 않고 js에서 필요한 만큼 복사하도록 변경하여 유지보수성 향상

fabbro1 committed:
fix: article 태그에서 section 태그로 보다 적적한 시맨틱 태그로 변경

fabbro1 committed:
feat: logManager 테스트 코드 추가

fabbro1 committed:
file: 테스트 파일 위치 변경

fabbro1 committed:
feat: CurrentBalanceManager 테스트 코드 추가

fabbro1 committed:
feat: 임시 테스트 추가

fabbro1 committed:
refactor: 템플릿을 외부에서 만들지 않고 각 자판기에서 만들도록 변경

fabbro1 committed:
fix: 키보드에서도 금액이 없을 부족할 경우 상품 각겨 노출되도록 수정
